lib/mongoid/document.rb

```diff
@@ -353,7 +353,57 @@ def instantiate(attrs = nil, selected_fields = nil)
       #
       # @since 1.0.0
       def _types
-        @_type ||= (descendants + [ self ]).uniq.map(&:to_s)
+        @_type ||= (descendants + [ self ] + (@_type_mappings || {}).keys + descendants.map{|d| d._types }.flatten).map(&:to_s).uniq
+      end
+
+      # Returns the class name to instantiate after appying
+      # override_inheritance_type_value.
+      #
+      # @param [ String ] type The descendant type name (actual Ruby class name
+      # or arbitrary alias name configured using
+      # override_inheritance_type_value.
+      # @returns [ String ] Ruby document type name as camelcased string
+      def __type(type)
+        klass = (@_type_mappings || {})[type]
+        descendants.each {|d| break if klass ||= d.__type(type) } unless klass
+        klass
+      end
+
+      # Returns the class name to instantiate after appying
+      # override_inheritance_type_value.
+      #
+      # @param [ String ] type The descendant type name (actual Ruby class name
+      # or arbitrary alias name configured using
+      # override_inheritance_type_value.
+      # @eturns [ Class ] Ruby document type to instantiate
+      def _type(type)
+        klass = (__type(type) || type).constantize
+      end
+
+      def map_type(klass, name)
+        @_type_mappings ||= {}
+        @_type_mappings[name] = klass.name
+      end
+
+      def _overridden_inheritance_type_value
+        @overridden_inheritance_type_value
+      end
+
+      # Change the stored type name for single table inheritance. The default
+      # is to use the Ruby class name but that can prove problematic when
+      # there are multiple programs accessing the database with different
+      # class names or source languages.
+      #
+      # @example Override the type name
+      #
+      # class HospitalWorker::Doctor < HospitalWorker::Base
+      #   override_inheritance_type_value 'MD'
+      #
+      # @param [ String ] name The name to store in the type field
+      def override_inheritance_type_value(name)
+        c_name = name.to_s.camelize
```

Comment:
Why does this method alter the value provided to it?

Reply:
I wanted to treat :alias, 'Alias' and 'alias' the same so I camelized them all before storing.

Reply:
Converting to string is OK but transforming the values I would say should be left to the application.

Reply:
I think that's a good point. Your way increases flexibility. I don't think it will affect anything. The only potential issue I see is that in factory.rb the _type value is camelized before searching for the type name. It might be necessary to skip that step.

Reply:
Or maybe to compare the camelized version passed to ::__type against a camelized version from ::_types.